# Intro to Python

Welcome to your first notebook on python! 

In this notebook you will start your journey of becoming a Pythonista and Data Scientist. If you're completely new to programming this is the right place to start. But even if you are already familiar with Python or some of its concepts it will be a good revision to refresh your knowledge. 

At the end of the notbook you will...
* know how to use python as a calculator.
* be familiar with pythons concept of variables. 
* have an overview over Python's data types.

## A brief little introduction...

Python was developed in the late 1980s by Guido van Rossum. It is an interpreted, high-level and general purpose programming language. In comparison to other programming languages Python strives for a simpler, less-cluttered syntax and grammar. Nevertheless it is one of the (if not to say the) preferred and most used language when it comes to Data Science. Its versatility and flexibility in combination with its easy to learn and clean syntax makes it a perfect language for programming beginners. 

Enough said, let's directly dive into the first lesson. 

## Numeric Operations

At its base level, Python is really just an awesome calculator that can do way more stuff than addition and subtraction. But, let's focus on that functionality for now.

All of the simple operations that you think should be available are available. Addition, subtraction, multiplication, division and exponentiation are all accessible via + , - , * , / and ** , respectively.

In [1]:
7 + 8

15

In [2]:
7 - 8

-1

In [3]:
7 * 8

56

In [4]:
7 / 8

0.875

In [5]:
7 ** 8

5764801


Perfekt. All of these operations output exactly what we think they would.

Besides those simple operators you can also find on any calculator Python offers two more arithmetic operators which might be new to you: // and %.
The double slash // is called floor division. All it does is perform division and truncate the result. So where 7 / 8 gave us 0.875, 7 // 8 cuts off after the 0. giving us 0.

In [6]:
7 // 8

0


The last operation that we will go over is the modular division operator, % (also called modulo). This operation is the sibling to //. As you can see in the following example the // gives us the integer number of times that 7 goes into 71. But, there is still a remainder of one. The way we get the remainder of integer division is with the % operator.

In [7]:
71 // 7   # floor division

10

In [8]:
71 % 7    # modulo 

1

You can change the order in which operations get performed by using parenthesis (), just as you can in algebra.

In [9]:
4 + 5 * 3

19

In [10]:
(4 + 5) * 3

27

At this point we also want to introduce some simple functions which you will encounter during your daily work as a Data Scientist. (Don't worry if you don't know what a function is. We will cover this in another notebook.)

`abs()`, `min()`, `max()` and `round()` are usefull littel funcitons which will make your life a lot easier. As their names already imply `abs()` will return the absolut value of a number, `min()` and `max()` will return the minimum respectively maximum of a bunch of numbers and `round()` will round a number to a given amount of decimals.

In [11]:
abs(3.14)

3.14

In [12]:
abs(-3.41)


3.41

In [13]:
min(3, 5, 1, 6, 8, 9)

1

In [14]:
max(3, 5, 1, 6, 8, 9)

9

In [15]:
round(5.23412, 2)   # you can specify the amount of decimals after the comma

5.23

**Questions:**

What do you think the results of the following computations will be?

1. 8.0 - 7
2. 8 * 0.25
3. 5 ** 2
4. 17 // 5
5. (4 + 5) / 6
6. (46 / 8) % 2
7. abs(-7.9)
8. round(6.293764, 3)

<details><summary>
Click here for the answers.
</summary>
1. 1.0 
2. 2.0
3. 25
4. 3
5. 1.5
6. 1.75
7. 7.9
8. 6.294
</details>

## Variables

One of the most powerful constructs in programming is the ability to store arbitrary values in what we call variables. You can think of variable assignment as giving a name to something so that it can be accessed later by different parts of your program.

In Python, variable assignment occurs with the `=` operator. To assign a value to a variable name (i.e. declare it), you simply put the variable name on the left side of the = and the value you want to associate with that variable name on the right side. Once this has happened, you can access the value in the variable simply by using it's name somewhere later in your code.

In [16]:
x = 5
y = 2

In [17]:
x

5

In [18]:
x - 4

1

In [19]:
x + y

7

The name you can give a variable can technically be any contiguous set of characters, but there are some conventions followed in Python and programming in general. Python follows a variable naming convention called snake case. To write something in snake case, simply use a _ anywhere you would use a space, and make sure every word is lower case. For example, `this_is_a_variable`. Giving variables good names makes your code more readable and therefore maintainable. There is a big difference between seeing a variable called `degrees` and one called `y`. You should strive to give your variables well-defined, succinct names.

There are of course cases where using less than descriptive variable names follows convention and are, therefore, just fine to use. A common example is the use of `i` to keep track of an index. Because of its prevalent usage for indexing, it is usually easy to understand what is happening in that context when all you see is the varible name `i`. Here, the lack of descriptiveness is okay. The important thing is that the code is **understandable**.


Note that we saw no output from either x = 5 or y = 2 above. This is because the return value that would have been printed as output was assigned to the variables x and y, respectively. This is why we had to view them in the next lines.

A large part of variables' power is the fact that they can change (vary, if you will). This allows us to use a single variable name to keep track of a specific thing throughout the life of a program. Remember how we assigned the value 5 to x above? The exact same syntax can be used to change the value stored in the variable.

Say we want to make the value of x five more than it currently is. All we need to do is have x be assigned the value that results from adding 5 to x.

In [20]:
x = x + 5
x

10

Notice how the first line above is formatted. Python knows that the = means variable assignment, so when it sees the first line it evaluates the right side of the equals and then puts that value in x, even though x is part of the calculation on the right side. x is now connected with this new value and the old value is gone.

Changing variables in this way occurs so commonly that there is a built-in shorthand for it. The result of the first line could have been achieved with x += 5. This syntactic sugar is available for all the simple operations +, -, *, /, **, and % that we covered earlier.

**Questions:**

Consider the following code:
```
x = 5
y = 8
x += y
y = x - 3
x -= y - 5
```

What are the values of x and y after each line?

<details><summary>
Click here to see the values for x and y after the last line.
</summary>
x = 8;  
y = 10
</details>

## Types

You successfully used python as a simple calculator and learned how to use variables! The last topic we will cover in this notebook are data types. 

Data type is an important concept in programming. Variables can store data in different types and different types can do different things.

We will cover some of Python's built in data types in the following section. 

> Maybe you have heard or read that an important characteristic of Python is that it is a **duck typed** language. What does this mean? The name duck comes from the classic "If it walks like a duck, and quacks like a duck, then it must be a duck" adage. As applied to our situation, it simply means that Python will determine what it thinks is the best type to call a variable when you use it, unless explicitly told otherwise.

### Numeric types

Python has a lot of different data types. You've already used some of the base numeric types, which are built into Python. All of those represent a very simple idea, numbers. Numbers can be either `ints`, short for integers, `floats`, short for floating point/decimal numbers, or `complex`, which contain real and imaginary parts stored as floats.

In [21]:
7   # int 

7

In [22]:
3.5   # float

3.5

In [23]:
complex(3, 6)   # complex

(3+6j)

To inspect what type Python thinks a numeric (or anything else) is, you can pass it to the type() function. Let's see what we get out when we pass numbers of various types to this function.

In [24]:
type(7)

int

In [25]:
type(3.5)

float

In [26]:
type(complex(3, 6))

complex

As you can see, Python assumes that a number with no decimal point is an `int`, those with a decimal point a `float`, and (surprise!) those from the complex() constructor as `complex`.


Frequently, these subtle differences wont matter too much. However, there will be times when this implementation detail will make you think that something will work, when really it won't. Knowing how to check the type of something will help you solve any of these potential problems.

### String

Besides numeric types Python has a special type for text. This data type called `string` or short `str` represents sequences of characters. A string can contain as many characters as you want. While other programming languages have a special type called `char` for single characters, in Python single characters are also designated as `string`. Strings can also be empty or contain numbers. 
To make it clear for Python that something is a string you can use double  `" "` or single `' '` quotes. All characters between an opening delimiter and a matching closing delimiter are part of the string.

In [27]:
"Welcome at neuefische!"

'Welcome at neuefische!'

In [28]:
'We are looking forward to meeting you all soon :)'

'We are looking forward to meeting you all soon :)'

In [29]:
type("We hope you are also excited.")

str

In [30]:
type('')

str

In [31]:
type("A")

str

In [32]:
type("8128")

str

### Bool



Python has a type of variable called `bool`. It has two possible values: `True` and `False`. 

> Note that the boolean values `True` and `False` begin with capital letters. Python is a case sensitive language. That means True and true are not the same!


In [33]:
True

True

In [34]:
False

False

In [35]:
type(True)

bool

In [36]:
type(False)

bool

Instead of putting `True` or `False` directly into our code, we usually get those boolean values from comparison operations, which can either evaluate to `True` or `False`. The following table gives an overview over the possible comparison operations and the operators used in Python. 

| Operator | Operation | Description |
|----|-----------|-------------|
| == | a == b    | a equal to b | 
| < | a < b    | a less than b | 
| <= | a <= b   | a less than or equal to b |
| > | a > b    | a greater than b  | 
| >= | a >= b  | a greater than or equal to b | 
| != | a != b | a not equal to b | 

In [37]:
1 == 1


True

In [38]:
7 < 5

False

In [39]:
3 >= 3

True

You will have to deal a lot with booleans and one characteristic is especially interesting from the Data Scienctist's perspective. 

If you convert `True` or `False` to an integer using `int()` you will always get 1 for `True` and 0 for `False`. This will come in handy in some situations. 

In [40]:
int(True)

1

In [41]:
int(False)

0

**Questions:**

Determine the type of the following:

1. " "
2. 231.54
3. True
4. 96
5. "Hello"
6. 4 + 3i
7. '203 + 45i'


<details><summary>
Click here for the answers.
</summary>
1. string 
2. float
3. bool
4. integer
5. string
6. complex
7. string
</details>
